pkg/expression/builtin_string.go

```diff
@@ -759,6 +759,7 @@ func (c *reverseFunctionClass) getFunction(ctx sessionctx.Context, args []Expres
 	if err := c.verifyArgs(args); err != nil {
 		return nil, err
 	}
+	bitTypeFlag := types.IsTypeBit(args[0].GetType())
 	bf, err := newBaseBuiltinFuncWithTp(ctx, c.funcName, args, types.ETString, types.ETString)
```

Comment:
I noticed that the returned collation should be `binary`. You can check with `--column-type-info` argument in `mysql` cli.

Run the SQLs in the issue, we'll get:

MySQL Result:

```
mysql> select reverse(col_5 ) as r1  from tl67c6ff48  order by r1;
Field   1:  `r1`
Catalog:    `def`
Database:   ``
Table:      ``
Org_table:  ``
Type:       VAR_STRING
Collation:  binary (63)
Length:     37
Max_length: 5
Decimals:   31
Flags:      BINARY 


+--------------+
| r1           |
+--------------+
| 0x00D061AE16 |
| 0x0C57A3DB04 |
| 0x1594231F10 |
| 0x2076E8B81B |
| 0x21B5B7FF0D |
| 0x32B80C4F16 |
| 0x35ABE56507 |
| 0x4325A9F11A |
| 0x44E5E45C16 |
| 0x46D4473E0F |
| 0x68CACA5E17 |
| 0x763ACC810D |
| 0x82C1DBF714 |
| 0x886DA4D305 |
| 0x8FE7949000 |
| 0xA5EADAE81C |
| 0xAABF39DA1D |
| 0xB1D477E613 |
| 0xBF41716512 |
| 0xC403AF550A |
| 0xD55CF06313 |
| 0xD64C364706 |
| 0xE585E2A514 |
| 0xEEDF84D20E |
+--------------+
24 rows in set (0.00 sec)
```

Your PR result:

```
mysql> select reverse(col_5 ) as r1  from tl67c6ff48  order by r1;
Field   1:  `r1`
Catalog:    `def`
Database:   ``
Table:      ``
Org_table:  ``
Type:       VAR_STRING
Collation:  utf8mb4_bin (46)
Length:     20
Max_length: 5
Decimals:   31
Flags:      


+-------+
| r1    |
+-------+
|  a   |
| 
  W   |
| #  |
|  v |
    |
| 2
   O  |
| 5e   |
| U
   |
| C%   |
| D\   |
| FG>  |
| h^   |
| m    |
|      |
  |:́
|      |
| 9    |
| 甐    |
| w    |
| \c   |
| L6G  |
| Aqe  |
|      |
| ߄    |
+-------+
24 rows in set (0.08 sec)
```

Maybe a better way to solve this issue is to use `binary` collation, then the issue is automatically solved (and we don't need to handle `bitTypeFlag` specially, as it'll be matched by `types.IsBinaryStr`. This issue not only affects `reverse`, it also affects other functions like `concat`: `select concat(col_5, '1') from tl67c6ff48;` will have `binary` collation on MySQL, and will be `utf8mb4_bin` on TiDB.

One fix of this problem is to set collation to `binary` in `inferCollation` like `JSON` type (which sets collation to `utf8mb4_bin`. For example:

```
	if types.IsTypeBit(exprs[0].GetType()) {
		dstCharset, dstCollation = charset.CharsetBin, charset.CollationBin
	}
...
...
		if types.IsTypeBit(arg.GetType()) {
			argCharset, argCharset = charset.CharsetBin, charset.CollationBin
		}
```

`deriveCoercibilityForColumn` also needs to be modified. It should return `CoercibilityImplicit` when the type is `BIT`. You can test/check this behavior with `select COERCIBILITY(...)`. After these two modification, the `reverse` and `concat` should work fine.

Reply:
> One fix of this problem is to set collation to `binary` in `inferCollation` like `JSON` type (which sets collation to `utf8mb4_bin`. For example:
> 
> ```
> 	if types.IsTypeBit(exprs[0].GetType()) {
> 		dstCharset, dstCollation = charset.CharsetBin, charset.CollationBin
> 	}
> ...
> ...
> 		if types.IsTypeBit(arg.GetType()) {
> 			argCharset, argCharset = charset.CharsetBin, charset.CollationBin
> 		}
> ```
> 
> `deriveCoercibilityForColumn` also needs to be modified. It should return `CoercibilityImplicit` when the type is `BIT`. You can test/check this behavior with `select COERCIBILITY(...)`. After these two modification, the `reverse` and `concat` should work fine.

With this, it will make the return column type showed by `--column-type-info` compatible with mysql.

Reply:
I updated as you suggestion by https://github.com/pingcap/tidb/pull/50146/commits/b166a84bbf878bff1a90d8769a370a477168ec58.

Reply:
> @YangKeao I tested it can't work if fixed it just as above.

You are right, the collation of the return type of `cast` function is still `utf8mb4_bin`. This PR looks good to me. Thank you 🍻 .